[SPARK-40416][SQL] Move subquery expression CheckAnalysis error messages to use the new error framework

dtenedor opened the pull request:
### What changes were proposed in this pull request?

Move subquery expression CheckAnalysis error messages to use the new error framework.

This will help improve the usability of Apache Spark as a product, and link against documentation.

### Why are the changes needed?

Error messages related to SQL query analysis in Apache Spark need some work to make them more descriptive and actionable.

### Does this PR introduce _any_ user-facing change?

Yes, error messages change.

### How was this patch tested?

Unit test and query test coverage shows the updates in error messages received.

dtenedor committed:
initial commit

dtenedor committed:
initial commit

dtenedor commented:
Hi @MaxGekk @srielau  can you please help review this? 🙏 it helps make error messages better with respect to subquery expression analysis. @allisonwang-db FYI

dtenedor committed:
update tests

dtenedor committed:
fix tests

allisonwang-db commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,83 @@
   },
+  "INVALID_SUBQUERY_EXPRESSION" : {
+    "message" : [
+      "Invalid subquery expression"
```
```suggestion
      "Invalid subquery: "
```

dtenedor replied:
Done

allisonwang-db commented on sql/core/src/test/resources/sql-tests/results/join-lateral.sql.out:
```diff
@@ -323,14 +322,10 @@ SELECT * FROM t1, LATERAL (SELECT rand(0) FROM t2)
 -- !query output
 org.apache.spark.sql.AnalysisException
-Non-deterministic lateral subqueries are not supported when joining with outer relations that produce more than one row
-SubqueryAlias __auto_generated_subquery_name
-+- Project [rand(0) AS rand(0)#x]
-   +- SubqueryAlias spark_catalog.default.t2
-      +- View (`spark_catalog`.`default`.`t2`, [c1#x,c2#x])
-         +- Project [cast(col1#x as int) AS c1#x, cast(col2#x as int) AS c2#x]
-            +- LocalRelation [col1#x, col2#x]
-; line 1 pos 9
```
The problematic query plan will not get printed out after this change?

dtenedor replied:
This is a good question. Let's talk offline about it in general a bit.

dtenedor replied:
Update on this: it looks like @gengliangwang  has this PR to support adding the origin information to `AnalysisException`: https://github.com/apache/spark/pull/37841. We can think about whether to include or elide the logical plan strings in those cases.

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -923,7 +966,11 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
-        failAnalysis(s"Accessing outer query column is not allowed in:\n$p")
+        throw new AnalysisException(
+          errorClass = "INVALID_SUBQUERY_EXPRESSION",
+          errorSubClass = "AGGREGATE_FUNCTION_MIXED_OUTER_LOCAL_REFERENCES",
```
ACCESSING_OUTER_QUERY_COLUMN_IS_NOT_ALLOWED

dtenedor replied:
Nice catch! Fixed

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -923,7 +966,11 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
+          errorClass = "INVALID_SUBQUERY_EXPRESSION",
+          errorSubClass = "AGGREGATE_FUNCTION_MIXED_OUTER_LOCAL_REFERENCES",
+          origin = p.origin,
+          messageParameters = Array.empty[String])
```
Should we include the query plan here? Otherwise, it's hard to tell what's wrong.

dtenedor replied:
Same as about, we include the `origin` here which indicates the location in the input SQL query at which the error took place. Let's talk about it offline though, we can go either way on this.

allisonwang-db commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,83 @@
+          "Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses"
+        ]
+      },
+      "IN_EXISTS_SUBQUERIES_ONLY_IN_FILTER_AGGREGATE_PROJECT" : {
```
This list might change in the future. We can make the name more generic.

dtenedor replied:
Good idea, done. Also made this fix for `CORRELATED_SCALAR_SUBQUERIES_ONLY_IN_FILTER_AGGREGATE_PROJECT` as well.

allisonwang-db commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,83 @@
+          "Lateral join condition cannot be non-deterministic: <condition>"
+        ]
+      },
+      "MORE_THAN_ONE_OUTPUT_COLUMN" : {
```
```suggestion
      "SCALAR_SUBQUERY_RETURN_MORE_THAN_ONE_OUTPUT_COLUMN" : {
```

dtenedor replied:
Done.

allisonwang-db commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,83 @@
+          "Scalar subquery must return only one column, but got <number>"
+        ]
+      },
+      "MUST_AGGREGATE_CORRELATED_SUBQUERY" : {
```
```suggestion
      "MUST_AGGREGATE_CORRELATED_SCALAR_SUBQUERY" : {
```

dtenedor replied:
Done.

allisonwang-db commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,83 @@
+          "Correlated scalar subqueries must be aggregated"
+        ]
+      },
+      "MUST_AGGREGATE_CORRELATED_SUBQUERY_OUTPUT" : {
```
This looks very similar to the one above. Maybe we can unify them later.

dtenedor replied:
Good idea. Let's think about it. For now I renamed it to `MUST_AGGREGATE_CORRELATED_SCALAR_SUBQUERY_OUTPUT` to keep consistent with your other suggestion.

allisonwang-db reviewed (commented):
This is awesome! Thanks @dtenedor.

dtenedor committed:
respond to code review comments

dtenedor committed:
remove unused code

dtenedor committed:
remove unused code

dtenedor committed:
fix tests

dtenedor committed:
fix tests

AmplabJenkins commented:
Can one of the admins verify this patch?

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -2005,13 +2018,17 @@ class SubquerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
           |""".stripMargin),
         Row(5) :: Row(null) :: Nil)
-      assert(intercept[AnalysisException] {
+      val msg1 = intercept[AnalysisException] {
```
It is an exception not a message.

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -1983,9 +1992,13 @@ class SubquerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       Seq((0, 1)).toDF("c1", "c2").createOrReplaceTempView("t1")
       Seq((1, 2), (2, 2)).toDF("c1", "c2").createOrReplaceTempView("t2")
-      assert(intercept[AnalysisException] {
+      val msg1 = intercept[AnalysisException] {
```
Please, rename the val. It is not a message.

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -523,30 +523,39 @@ class SubquerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
   }
 
   test("non-equal correlated scalar subquery") {
     val msg1 = intercept[AnalysisException] {
```
ditto

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -523,30 +523,39 @@ class SubquerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
   }
 
   test("non-aggregated correlated scalar subquery") {
     val msg1 = intercept[AnalysisException] {
```
ditto

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -523,30 +523,39 @@ class SubquerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       val errMsg = intercept[AnalysisException] {
```
ditto

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,83 @@
+          "Correlated column is not allowed in predicate"
+        ]
+      },
+      "CORRELATED_SCALAR_SUBQUERIES_IN_GROUP_BY_MUST_BE_IN_AGGREGATE_EXPRESSIONS" : {
```
Could you replace it by a shorter one, please. The error classes are placed to doc headers. The long error classes can cause some issues in formatting. cc @srielau

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -743,7 +743,11 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
       })
       if (aggregates.isEmpty) {
-        failAnalysis("The output of a correlated scalar subquery must be aggregated")
+        throw new AnalysisException(
```
The current approach is to raise exception from `Query*Errors`. Could move this to `QueryCompilationErrors`, please.

dtenedor committed:
Merge branch 'master' into subquery-error-classes
fetch latest from master

dtenedor committed:
respond to code review comments

stash

respond to code review comments

stash

respond to code review comments

update tests

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,78 @@
+    "subClass" : {
+      "ACCESSING_OUTER_QUERY_COLUMN_IS_NOT_ALLOWED" : {
+        "message" : [
+          "Accessing outer query column is not allowed in this location<plan>"
```
Yeah what we need here is the context, not the plan. Actually we need two locations. The place it's resolving to and the place it's resolving from.

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,78 @@
+      },
+      "CORRELATED_COLUMN_IS_NOT_ALLOWED_IN_PREDICATE" : {
+        "message" : [
+          "Correlated column is not allowed in predicate<plan>"
```
Context and column reference is needed here. Not plan

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,78 @@
+          "Correlated column is not allowed in predicate<plan>"
+        ]
+      },
+      "CORRELATED_SCALAR_SUBQUERIES_ONLY_IN_CERTAIN_PLACES" : {
```
Can we make this shorter?
UNSUPPORTED_CORRELATED_SCALAR_SUBQUERY maybe

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,78 @@
+          "Correlated scalar subqueries can only be used in filters, aggregations, projections, and UPDATE/MERGE/DELETE commands<plan>"
+        ]
+      },
+      "IN_EXISTS_SUBQUERIES_ONLY_CERTAIN_PLACES" : {
```
How about UNSUPPORTED_IN_EXIST_SUBQUERY

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,78 @@
+          "Non-deterministic lateral subqueries are not supported when joining with outer relations that produce more than one row<plan>"
+        ]
+      },
+      "OUTER_ATTRIBUTE_NOT_FOUND" : {
```
What is an attribute? Is this an internal error? Or is this jut an UNRESOLVED_COLUMN?

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,78 @@
+          "Outer attribute not found: <value>"
+        ]
+      },
+      "OUTER_REFERENCES_ONLY_ALLOWED_IN_WHERE_HAVING" : {
```
What is the difference between OUTER and CORRELATED? Should this be CORRELATE_REFERENCE?
Also, as we lift restrictions, maybe we should not include the "whitelist" in the error class name,
How about:
UNSUPPORTED_CORRELATED_REFERENCE.
Then whitelist  in the error test and we can loosen it up over time.

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -327,6 +327,78 @@
+          "Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses<plan>"
+        ]
+      },
+      "SCALAR_SUBQUERY_RETURN_MORE_THAN_ONE_OUTPUT_COLUMN" : {
```
Hmm, this one is different. It's a proper semantic error and should have it' own error class IMHO. 
Most/all of the others are product limitations.

srielau commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -756,10 +756,8 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
-            "cannot contain non-correlated columns: " +
-            invalidCols.mkString(","))
+        throw QueryCompilationErrors.nonCorrelatedColumnsInGroupBy(
+          invalidCols.map(_.name).mkString(","), expr.origin)
```
@gengliangwang is working on the origin. Link up with him on whether you should leave it out here

gengliangwang replied:
With https://github.com/apache/spark/pull/37841 merged, we can skip passing expr.origin here. @dtenedor Will you give a try?

srielau commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -1543,10 +1543,108 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+      errorClass = "INVALID_SUBQUERY_EXPRESSION",
+      errorSubClass = "ACCESSING_OUTER_QUERY_COLUMN_IS_NOT_ALLOWED",
+      origin = origin,
+      messageParameters = Array(if (SQLConf.get.includePlansInErrors) s": $plan" else ""))
```
General comment: Please use Map to guard against ordering errors: Map("plan" -> ....)
(my comment on removing <plan> not withstanding)

dtenedor reviewed (commented):
Thank you all @allisonwang-db @srielau @MaxGekk @gengliangwang for your helpful review. I responded to all comments, please take another look. Hat-tip to Gengliang with a helpful suggestion to add an overload `expression.failAnalysis` to propagate query contexts while adding the new error classes and minimizing diffs in `CheckAnalysis`.

dtenedor committed:
respond to code review comments

dtenedor committed:
fix tests

dtenedor committed:
Merge branch 'master' into subquery-error-classes
fetch latest from master

dtenedor committed:
fix test

dtenedor committed:
fetch latest from master

allisonwang-db commented on sql/core/src/test/resources/sql-tests/results/join-lateral.sql.out:
```diff
@@ -317,11 +317,18 @@ SELECT * FROM t1, LATERAL (SELECT c1 + c2 + rand(0) AS c3)
 struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
-Non-deterministic lateral subqueries are not supported when joining with outer relations that produce more than one row
```
Should we set `spark.sql.error.includePlans` to `true` when generating this golden file?

allisonwang-db replied:
Maybe we should output the canonicalized plan string instead of using `toString`?

dtenedor replied:
Hi @allisonwang-db  I looked at this and the existing `plan.canonicalized` uses `transformUp` which unfortunately is not allowed during analysis. However I was able to implement this by manually canonicalizing just the expression IDs for purposes of these strings and now the tests should be deterministic. Thanks for the suggestion.

dtenedor commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -695,6 +707,73 @@
   },
+  "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY" : {
+    "message" : [
+      "Invalid subquery expression:"
```
Done

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

update

update

update

update

dtenedor committed:
fix test

gengliangwang commented:
Hi @dtenedor, I have just merged the bug fix https://github.com/apache/spark/pull/37861. The query context should be set correctly in the AnlysisException.  I believe the test cases in this PR need to be updated so please try updating the code base to the latest master.

dtenedor committed:
Merge branch 'master' into subquery-error-classes

fetch latest changes from master

MaxGekk commented on sql/core/src/test/resources/sql-tests/results/udf/udf-except.sql.out:
```diff
@@ -100,12 +100,17 @@ WHERE  udf(t1.v) >= (SELECT   min(udf(t2.v))
+  "errorClass" : "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY",
+  "errorSubClass" : "CORRELATED_COLUMN_IS_NOT_ALLOWED_IN_PREDICATE",
+  "messageParameters" : {
+    "planString" : ": (cast(udf(cast(k#0 as string)) as string) = cast(udf(cast(outer(k#1) as string)) as string))"
```
Is it a plan string? Let's rename it.

dtenedor replied:
Good point, this comes from expressions rather than plans. Renamed all to `treeNode` instead of `planString`.

MaxGekk commented on sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out:
```diff
@@ -105,14 +131,20 @@ WHERE  t1a IN (SELECT t2a
+  "errorClass" : "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY",
+  "errorSubClass" : "UNSUPPORTED_CORRELATED_REFERENCE",
+  "messageParameters" : {
+    "planString" : ": Aggregate [min(outer(t2a#6)) AS min(outer(t2.t2a))#7]\n+- SubqueryAlias t3\n   +- View (`t3`, [t3a#3,t3b#4,t3c#5])\n      +- Project [cast(t3a#0 as int) AS t3a#3, cast(t3b#1 as int) AS t3b#4, cast(t3c#2 as int) AS t3c#5]\n         +- Project [t3a#0, t3b#1, t3c#2]\n            +- SubqueryAlias t3\n               +- LocalRelation [t3a#0, t3b#1, t3c#2]\n"
```
> t2a#6

Are the IDs stable? Should we replace them like
https://github.com/apache/spark/blob/db5aea60e4b22449d529f9e866366cfdc784b140/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala#L94

dtenedor replied:
Good Q. I found they are not stable when running the tests outside of my local machine, therefore to prevent the tests from becoming flaky, I found it necessary to normalize the expression IDs. I added code in the `failAnalysis` overload that takes `TreeNode`s to do that, iterating through the trees to reset the expression IDs. This should be more robust than regex-replacing the string. We do have the existing `canonicalized` method on logical plans and expressions, but that uses `transformUp` which is not allowed in the analyzer, so I ended up doing it separately here for this purpose.

dtenedor committed:
reuse existing plan canonicalize method

respond to code review comments

update

update

update

simplify

reuse existing plan canonicalize method

respond to code review comments

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -1563,10 +1563,12 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+      errorClass = "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY",
+      errorSubClass = "AGGREGATE_FUNCTION_MIXED_OUTER_LOCAL_REFERENCES",
+      origin = origin,
+      messageParameters = Map("function" -> funcStr))
```
Should we quote the function? by `toSQLExpr` if it is an expression or maybe by `toSQLId`? cc @srielau

dtenedor replied:
Good Q,
Looks like this is a `sql` string generated from a predicate which may contain multiple functions. It renders from the error-classes.json file like this:

```
Found an aggregate function in a correlated predicate that has both outer and local references, which is not supported: <function>
```

It might be better to leave it unquoted since some parts of the predicate may contain their own double-quotes, e.g. literal string values. Will leave it alone for now, Serge please comment if you want it to work differently and I can update it.

dtenedor commented:
> Hi @dtenedor, I have just merged the bug fix https://github.com/apache/spark/pull/37861. The query context should be set correctly in the AnlysisException. I believe the test cases in this PR need to be updated so please try updating the code base to the latest master.

@gengliangwang thanks, this is done. We can see the new query contexts in this PR at the latest commit, they appear more accurate now.

dtenedor commented:
Hi @MaxGekk I responded to your comments, please take another look at this when you have time. 🙏 @gengliangwang @allisonwang-db @srielau FYI

gengliangwang commented on sql/core/src/test/resources/sql-tests/results/join-lateral.sql.out:
```diff
@@ -317,11 +317,20 @@ SELECT * FROM t1, LATERAL (SELECT c1 + c2 + rand(0) AS c3)
+  "errorClass" : "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY",
+  "errorSubClass" : "NON_DETERMINISTIC_LATERAL_SUBQUERIES",
+  "messageParameters" : {
+    "treeNode" : ": !LateralJoin lateral-subquery#5 [c1#6 && c2#7], Inner\n:  +- SubqueryAlias __auto_generated_subquery_name\n:     +- Project [(cast((outer(c1#2) + outer(c2#3)) as double) + rand(0)) AS c3#4]\n:        +- OneRowRelation\n+- SubqueryAlias spark_catalog.default.t1\n   +- View (`spark_catalog`.`default`.`t1`, [c1#2,c2#3])\n      +- Project [cast(col1#0 as int) AS c1#2, cast(col2#1  ...
```
Given there is already query context in the error message, are we sure about showing the logical plan as well? (This PR contains a lot of threads, I am not sure if this is discussed.)
cc @srielau

dtenedor replied:
Yes we did discuss this already in the PR threads :) I suggested eliding the plan strings given that we have the query context now, but some folks opined that the strings are helpful for debugging. In the end I stuck a compromise by including the strings but covered by a new SQLConf to toggle their presence. This new config can remain enabled by default for now in Apache Spark so folks who are used to the strings for debugging may continue to use them.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -823,12 +834,16 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
+                a.failAnalysis(
+                  errorClass = "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY",
+                  errorSubClass = "MUST_AGGREGATE_CORRELATED_SCALAR_SUBQUERY",
+                  treeNodes = Seq(a))
```
why do we need to pass `a` itself?

dtenedor replied:
Good Q. The way I implemented it, `a` is a `TreeNode`. This overload of `failAnalysis` I added takes a `Seq[TreeNode[_]]` and then assigns their `toString`s to the `treeNode` parameter. I do this in the `failAnalysis` overload itself in order to normalize the expression IDs in the string to keep tests deterministic (the LogicalPlan `canonicalized` method aims to support this as well, but is not available during analysis since it uses `transformUp`).

dtenedor commented:
Friendly ping @MaxGekk @allisonwang-db @gengliangwang can we please review this again?

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/package.scala:
```diff
@@ -45,18 +51,101 @@ package object analysis {
+        errorClass: String,
+        errorSubClass: String,
+        treeNodes: Seq[TreeNode[_]]): Nothing = {
+      // Normalize expression IDs in the query plan to keep tests deterministic.
```
Shall we use `Expression.canonicalized` instead? There are existing methods for such purposes.

dtenedor replied:
I tried this, but this `canonicalized` method is not available during analysis since it uses `transformUp`. Therefore a separate implementation is needed.

dtenedor replied:
![image](https://user-images.githubusercontent.com/99207096/190524824-372496a8-7a6f-464d-9239-57df10ecd168.png)

gengliangwang replied:
Well, at least I would suggest moving these methods out of this file. We can do it in `Expression` or other Util objects.

dtenedor replied:
Hi @gengliangwang I looked at this at length some more, and I was able to reuse the existing plan `canonicalize` method after all using `AnalysisHelper.allowInvokingTransformsInAnalyzer` in one specific place instead. Now there is no need to introduce new code for this purpose. Please take another look.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -730,6 +729,13 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
   }
 
+  private def canonicalizeForError(expr: LogicalPlan): String =
+    if (SQLConf.get.getConf(SQLConf.CANONICALIZE_PLANS_IN_ERRORS)) {
```
Actually we don't need such a new config. We can just check 
```
Utils.isTesting
```

dtenedor replied:
@gengliangwang thanks for the suggestion! This is done, it is simpler now.

gengliangwang commented:
LGTM except one comment

dtenedor committed:
fix test again

dtenedor committed:
fix test again

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -3911,6 +3911,15 @@ object SQLConf {
     .checkValues(ErrorMessageFormat.values.map(_.toString))
     .createWithDefault(ErrorMessageFormat.PRETTY.toString)
 
+  val INCLUDE_PLANS_IN_ERRORS = buildConf("spark.sql.error.includePlans")
```
@dtenedor let's revert the changes in SQLConf.scala

dtenedor replied:
@gengliangwang apologies for missing that. It is reverted now.

dtenedor committed:
respond to code review comments

gengliangwang commented:
Thanks, merging to master
